### PR TITLE
[#50] Support for AutoGenerationOption.

### PR DESCRIPTION
**Pull request by samuelmale**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionsCsvParser.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		super(lineProcessor);
+		service = identifierSourceService;
+	}
+	
+	@Override
+	public AutoGenerationOption bootstrap(CsvLine line) throws IllegalArgumentException {
+		AutoGenerationOption option = null;
+		String uuid = line.getUuid();
+		if (StringUtils.isNotBlank(uuid)) {
+			option = service.getAutoGenerationOptionByUuid(uuid);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionsCsvParser.java:30]: This requires [IDGEN-114](https://issues.openmrs.org/browse/IDGEN-114) to be in the box, so you will have to do this eventually **for all beans connected to this domain**:
```java
@OpenmrsProfile(modules = { "idgen:4.6.0-SNAPSHOT" })
```
(Snapshot versions should be handled fine, see [here](https://github.com/openmrs/openmrs-core/blob/750428347aed88a7f34d2ee98f66e9d436f57425/api/src/main/java/org/openmrs/module/ModuleUtil.java#L427-L436).)

Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionsCsvParser.java:30]: @mks-d not sure why, `@OpenmrsProfile(modules = { "idgen:4.6.0-SNAPSHOT" })` doesn't work. Spring fails to load the application context. However, `@OpenmrsProfile(modules = { "idgen:*" })` works out of the box.

File: api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionLineProcessor.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	final public static String AUTO_GEN_ENABLED = "auto generation enabled";
+	
+	@Autowired
+	private LocationService locationService;
+	
+	@Autowired
+	private IdentifierSourceService identifierSourceService;
+	
+	@Autowired
+	private PatientService patientService;
```
[the remark this answers is not in the document]
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionLineProcessor.java]: Resolved

File: api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionLineProcessor.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+			IdentifierSource source = identifierSourceService.getIdentifierSourceByUuid(identifierSourceRef);
+			generationOption.setSource(source);
+		}
+		// Fill location
+		{
+			String locationRef = line.get(LOCATION);
+			Location location = locationService.getLocationByUuid(locationRef);
+			if (location == null) {
+				location = locationService.getLocation(locationRef);
+			}
```
[the remark this answers is not in the document]
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionLineProcessor.java]: Resolved

File: api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionsCsvParser.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import org.openmrs.module.initializer.api.CsvParser;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@OpenmrsProfile(modules = { "idgen:4.6.0-SNAPSHOT" })
+public class AutoGenerationOptionsCsvParser extends CsvParser<AutoGenerationOption, BaseLineProcessor<AutoGenerationOption>> {
+	
+	private IdentifierSourceService service;
+	
+	@Autowired
+	protected AutoGenerationOptionsCsvParser(IdentifierSourceService identifierSourceService,
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionsCsvParser.java:19]: Shouldn't this be `public`?

File: api/src/test/java/org/openmrs/module/initializer/api/AutoGenerationOptionsLoaderIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import org.openmrs.module.initializer.DomainBaseModuleContextSensitiveTest;
+import org.openmrs.module.initializer.api.idgen.autogen.AutoGenerationOptionsLoader;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import static org.hamcrest.Matchers.is;
+
+public class AutoGenerationOptionsLoaderIntegrationTest extends DomainBaseModuleContextSensitiveTest {
+	
+	@Autowired
+	AutoGenerationOptionsLoader loader;
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/AutoGenerationOptionsLoaderIntegrationTest.java:17]: Make this `private`, it's just cleaner.

File: api/src/test/java/org/openmrs/module/initializer/api/AutoGenerationOptionsLoaderIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+		option = identifierSourceService.getAutoGenerationOptionByUuid("ca5cc0bf-38f0-41a6-9759-dd2a6763155a");
+		Assert.assertNotNull(option);
+		Assert.assertTrue(option.isAutomaticGenerationEnabled());
+		Assert.assertFalse(option.isManualEntryEnabled());
+		
+		//Verify edited
+		option = identifierSourceService.getAutoGenerationOptionByUuid(OPTION_TO_EDIT_UUID);
+		Assert.assertFalse(option.isAutomaticGenerationEnabled());
+		Assert.assertTrue(option.isManualEntryEnabled());
+		Assert.assertThat(option.getIdentifierType().getName(), is("Legacy ID"));
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/AutoGenerationOptionsLoaderIntegrationTest.java:47]: Can they be retired, if yes, can you add a case for it?

File: api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionLineProcessor.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	@Autowired
+	public AutoGenerationOptionLineProcessor(@Qualifier("locationService") LocationService ls, IdentifierSourceService iss,
+	    @Qualifier("patientService") PatientService ps) {
+		this.ls = ls;
+		this.iss = iss;
+		this.ps = ps;
+	}
+	
+	@Override
+	public AutoGenerationOption fill(AutoGenerationOption generationOption, CsvLine line) throws IllegalArgumentException {
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionLineProcessor.java:44]: `generationOption` → `option`

File: api/src/test/resources/testAppDataDir/configuration/autogenerationoptions/autogenerationoptions.csv [hunk 8]
```diff
@@ -0,0 +1,3 @@
+Uuid,Identifier Type,Location,Identifier Source,Manual Entry Enabled,Auto Generation Enabled,Sort Weight,_order:2000
+eade77b6-3365-47ed-9ee3-2324598629eb,Legacy ID,f0f8ba64b-ea57-4a41-b33c-9dfc59b0c60a,9eca4f4e-707f-4lb8-8289-2f9b6e93803f,true,false,,
+ca5cc0bf-38f0-41a6-9759-dd2a6763155a,OpenMRS ID,,0d47284f-9e9b-4a81-a88b-8bb42bc0a901,false,true,,
```
Comment (mks-d) [inline api/src/test/resources/testAppDataDir/configuration/autogenerationoptions/autogenerationoptions.csv:3]: Something's wrong with the CSV, thereby demonstrating that tests are inadequate, look at the 'Location' column.

Can you setup tests that work with the location then?

File: readme/autogenerationoptions.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+create new autogeneration options. Here is a sample CSV:
+
+| <sub>Uuid</sub>                                 | <sub>Identifier Type</sub> | <sub>Location</sub>      | <sub>Identifier Source</sub>                    | <sub>Manual Entry Enabled</sub> | <sub>Auto Generation Enabled</sub> | <sub>Sort Weight</sub> | <sub>_order:2000</sub> |
+|--------------------------------------|-----------------|---------------|--------------------------------------|----------------------|-------------------------|-------------|-------------|
+| eade77b6-3365-47ed-9ee3-2324598629eb | Legacy ID       |  | 9eca4f4e-707f-4lb8-8289-2f9b6e93803f | true                 | false                   |             |             |
+
+
+
+###### Header `Identifier Type`
+The **patient identifier type** associated with the autogeneration option. This is a mandatory property.
```
Comment (mks-d) [inline readme/autogenerationoptions.md:23]: This is not how mandatory fields are documented, look [here](https://github.com/mekomsolutions/openmrs-module-initializer/blob/bb5240a07781e79a3fd7436c362a36af0ade5a91/readme/mdm.md#header-metadata-set-uuid-mandatory) for instance.

When a column value is a reference to another entity, explain in details how the value can be set. Is this UUID only? If not, what else can be done?

File: api/src/test/resources/testAppDataDir/configuration/autogenerationoptions/autogenerationoptions.csv (hunk 8 above)
Comment (mks-d) [reply api/src/test/resources/testAppDataDir/configuration/autogenerationoptions/autogenerationoptions.csv:3]: @samuelmale sorry, somehow I mis-copy/pasted the CSV (Google Sheets failed on me), you can disregard [this](https://github.com/mekomsolutions/openmrs-module-initializer/pull/54#discussion_r408088918) comment.

Comment (samuelmale) [reply api/src/test/resources/testAppDataDir/configuration/autogenerationoptions/autogenerationoptions.csv:3]: There is nothing wrong with the Location column in the above CSV file, It's just that the Location column is optional.

Line one has a value: 'f0f8ba64b-ea57-4a41-b33c-9dfc59b0c60a'
Line two has no value.

File: api/src/test/java/org/openmrs/module/initializer/api/AutoGenerationOptionsLoaderIntegrationTest.java (hunk 6 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/initializer/api/AutoGenerationOptionsLoaderIntegrationTest.java:47]: They aren't retire-able

File: api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionLineProcessor.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+		
+		// Fill identifier type
+		{
+			String identifierTypeRef = line.get(IDENTIFIER_TYPE, true);
+			option.setIdentifierType(Utils.fetchPatientIdentifierType(identifierTypeRef, ps));
+		}
+		// Fill identifier source
+		{
+			String identifierSourceRef = line.get(IDENTIFIER_SOURCE, true);
+			IdentifierSource source = iss.getIdentifierSourceByUuid(identifierSourceRef);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionLineProcessor.java:57]: Can `IdentifierSource`s be fetched by name? If yes, please introduce a fetcher method that can handle both name and UUID.

Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/initializer/api/idgen/autogen/AutoGenerationOptionLineProcessor.java:57]: No, this is not handled in the service layer.

**Comment by samuelmale** [comment]
> 'Sort Weight' is not documented/used... etc, what's up with that one?

My bad, I wonder why I added the property in my past commits. Let me remove the property @mks-d

**Pull request merged**